source/2_platform/rex_directx/src/directx_renderer.cpp

```diff
@@ -73,76 +238,1816 @@ namespace rex
+        g_ctx.resource_pool.insert(resourceSlot, rsl::make_unique<InputLayoutResource>(input_element_descriptions));
+
+        return true;
+      }
+
+      //-------------------------------------------------------------------------
+      bool create_vertex_buffer(const commands::CreateBufferCommandDesc& cb, const ResourceSlot& resourceSlot)
+      {
+        REX_ASSERT_X(cb.buffer_data.data() != nullptr && cb.buffer_data.size() != 0, "Trying to create an empty vertex buffer");
+
+        rsl::unique_ptr<BufferResource> resource = internal::create_buffer((void*)cb.buffer_data.data(), rsl::safe_numeric_cast<s32>((cb.buffer_data.size().size_in_bytes())));
+
+        if(resource == nullptr)
+        {
+          REX_ERROR(LogDirectX, "Failed to create vertex buffer");
+          return false;
+        }
+
+        g_ctx.resource_pool.insert(resourceSlot, rsl::move(resource));
+
+        return true;
+      }
+
+      //-------------------------------------------------------------------------
+      bool create_index_buffer(const commands::CreateBufferCommandDesc& cb, const ResourceSlot& resourceSlot)
+      {
+        REX_ASSERT_X(cb.buffer_data.data() != nullptr && cb.buffer_data.size() != 0, "Trying to create an empty index buffer");
+
+        rsl::unique_ptr<BufferResource> resource = internal::create_buffer((void*)cb.buffer_data.data(), rsl::safe_numeric_cast<s32>(cb.buffer_data.size().size_in_bytes()));
+
+        if(resource == nullptr)
+        {
+          REX_ERROR(LogDirectX, "Failed to create index buffer");
+          return false;
+        }
+
+        g_ctx.resource_pool.insert(resourceSlot, rsl::move(resource));
+
+        return true;
+      }
+
+      //-------------------------------------------------------------------------
+      bool create_constant_buffer_view(const commands::CreateConstantBufferViewCommandDesc& cb, const ResourceSlot& resourceSlot)
+      {
+        REX_ASSERT_X(cb.buffer_size != 0, "Trying to create an empty constant buffer"); // cb.data is allowed to be NULL when creating a constant buffer
+
+        rsl::unique_ptr<ConstantBufferViewResource> resource = internal::create_constant_buffer_view(cb.frame_slot, cb.commited_resource, cb.buffer_size);
+
+        if(resource == nullptr)
+        {
+          REX_ERROR(LogDirectX, "Failed to create constant buffer");
+          return false;
+        }
+
+        g_ctx.resource_pool.insert(resourceSlot, rsl::move(resource));
+
+        return true;
+      }
+
+      //-------------------------------------------------------------------------
+      bool create_pipeline_state_object(const commands::CreatePipelineStateCommandDesc& cps, const ResourceSlot& resourceSlot)
+      {
+        REX_ASSERT_X(cps.input_layout.is_valid(), "Invalid input layout resource slot given");
+        REX_ASSERT_X(cps.shader_program.is_valid(), "Invalid shader program resource slot given");
+
+        rsl::hash_result hash = rsl::hash<commands::CreatePipelineStateCommandDesc> {}(cps);
```

Comment:
you don't need to hash things yourself, that defeats the purpose of an unordered_map

Reply:
rsl::unordered_set is not implemented